src/riscv/lib/src/jit/builder/instr_map.rs

```diff
@@ -0,0 +1,144 @@
+use std::fmt::Debug;
+
+use crate::jit::builder::instruction::LoweredInstruction;
+use crate::jit::builder::sequence::sequence_analysis::InstrInfo;
+use crate::machine_state::memory::Address;
+
+/// A holder of the Addresses of the instructions in a sequence.
+#[derive(Debug, Clone)]
+pub(super) struct AddrMap {
+    addresses: Vec<Address>,
+}
+
+impl AddrMap {
+    /// Create a new, empty AddrMap.
+    pub(super) fn new() -> Self {
+        Self {
+            addresses: Vec::new(),
+        }
+    }
+
+    /// Insert a new address into the map.
+    pub(super) fn insert(&mut self, address: Address) {
+        self.addresses.push(address);
+    }
+
+    /// Retrieve the address at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> Address {
+        self.addresses[instr_id.0]
+    }
+
+    /// Get the instr ID of the instruction at the given address.
+    pub(super) fn get_instr_id(&self, address: Address) -> Option<InstrId> {
+        self.addresses.binary_search(&address).ok().map(InstrId)
+    }
+}
+
+/// A holder of the instructions in a sequence with quick lookup via instr ID or address.
+#[derive(Debug, Clone)]
+pub(super) struct InstrMap<V> {
+    instructions: Vec<V>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct InstrId(usize);
+
+impl<V> InstrMap<V> {
+    /// Retrieve the instruction at the provided instr ID.
+    pub(super) fn get(&self, instr_id: InstrId) -> &V {
+        &self
+            .instructions
+            .get(instr_id.0)
+            .expect("instrID out of bounds.")
+    }
+
+    /// Retrieve a mutable reference to the instruction at the provided instr ID.
+    pub(super) fn get_mut(&mut self, instr_id: InstrId) -> &mut V {
+        self.instructions
+            .get_mut(instr_id.0)
+            .expect("instrID out of bounds.")
+    }
+
+    pub(super) fn iter(&self) -> impl Iterator<Item = (InstrId, &V)> {
+        self.instructions
+            .iter()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    pub(super) fn iter_mut(&mut self) -> impl Iterator<Item = (InstrId, &mut V)> {
+        self.instructions
+            .iter_mut()
+            .enumerate()
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    pub(super) fn get_first_mut(&mut self) -> (InstrId, &mut V) {
+        let idx = 0;
+        (
+            InstrId(idx),
+            self.instructions
+                .get_mut(idx)
+                .expect("Getting first instruction failed: No instructions in map."),
+        )
+    }
+
+    pub(super) fn get_last_mut(&mut self) -> (InstrId, &mut V) {
+        let idx = self.instructions.len() - 1;
+        (
+            InstrId(idx),
+            self.instructions
+                .get_mut(idx)
+                .expect("Getting last instruction failed: No instructions in map."),
+        )
+    }
+
+    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(&V) -> U,
```

Comment:
When you traverse, you want the instruction ID in some cases.

```suggestion
    pub(super) fn map<F, U>(self, f: F) -> InstrMap<U>
    where
        F: Fn(InstrId, &V) -> U,
```

Reply:
done [here](https://github.com/tezos/riscv-pvm/compare/4fa38ba56933ca86bbf7c4e4ac24198048d61b09..85b992e5fd437cfb1875b15fce5e3424395aa173)